service/sts/provider.go

```diff
@@ -0,0 +1,118 @@
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"time"
+)
+
+// Provider retrieves temporary credentials from the STS service, and keeps track of
+// their expiration time. This provider must be used explicitly, as it is not
+// included in the credentials chain.
+//
+// Example how to configure a service to use this provider:
+//
+//		config := &aws.Config{
+//			Credentials: sts.NewCredentials(awsRegion, awsArn, 10*time.Second),
+//		})
+//		// Then use config when creating your AWS service using this SDK.
+//
+// Example how to configure the Provider with custom duration or session name.
+//
+//		p := &credentials.Provider{
+//			// Extend the duration to 1 hour
+//			Duration: time.Hour
+//			// Use custom role name
+//			RoleSessionName: "custom-session-name"
+//		}
+//
+type Provider struct {
+	// Regio to be used for the STS request.
+	Region string
+
+	// Role to be assumed.
+	RoleARN string
+
+	// Session name, if you wish to reuse the credentials elsewhere.
+	RoleSessionName string
+
+	// Expiry duration of the STS credentials.
+	Duration time.Duration
+
+	// ExpiryWindow will allow the credentials to trigger refreshing prior to
+	// the credentials actually expiring. This is beneficial so race conditions
+	// with expiring credentials do not cause request to fail unexpectedly
+	// due to ExpiredTokenException exceptions.
+	//
+	// So a ExpiryWindow of 10s would cause calls to IsExpired() to return true
+	// 10 seconds before the credentials are actually expired.
+	//
+	// If ExpiryWindow is 0 or less it will be ignored.
+	ExpiryWindow time.Duration
+
+	// The time of the credential expiry.
+	expiration time.Time
+}
+
+// NewCredentials returns a pointer to a new Credentials object wrapping the Provider.
+// The credentials will expire every 15 minutes and the role will be named after a nanosecond
+// timestamp of this operation.
+//
+// Region and roleARN parameters are used for building the STS "AssumeRole" call.
+//
+// Window is the expiry window that will be subtracted from the expiry returned
+// by the role credential request. This is done so that the credentials will
+// expire sooner than their actual lifespan.
+func NewCredentials(region, roleARN string, window time.Duration) *credentials.Credentials {
+	now := time.Now()
+
+	return credentials.NewCredentials(&Provider{
+		Region:  region,
+		RoleARN: roleARN,
+		// Expire as often as AWS permits.
+		Duration: 15 * time.Minute,
+		// Construct a role session name that should be unique.
+		RoleSessionName: fmt.Sprintf("%d", now.UTC().UnixNano()),
+		// Expire 10 seconds before the actual expiry to prevent exceptions.
+		ExpiryWindow: window,
+		// Force immediate refresh.
+		expiration: now,
+	})
+}
+
+// Retrieve generates a new set of temporary credentials using STS.
+func (s *Provider) Retrieve() (credentials.Value, error) {
+
+	stsClient := New(&aws.Config{
+		Region: s.Region,
+	})
+
+	roleOutput, err := stsClient.AssumeRole(&AssumeRoleInput{
```

Comment:
There are a few different ways STS can provide roles, AssumeRole and GetSessionToken.  Both are very similar with the exception of the RoleARN. It would be great if the single Provider could be built for both types cleanly. but if not it may make sense for this PR code to be just a AssumeRoleProvider instead of generic STS provider.